napari/utils/_tests/test_magicgui.py

```diff
@@ -66,6 +67,45 @@ def add_data() -> dtype:
     assert viewer.layers[0].source.widget == add_data
 
 
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason='Futures not subscriptable before py3.9'
+)
+@pytest.mark.parametrize('LayerType, data, ndim', test_data)
+def test_magicgui_add_future_data(make_napari_viewer, LayerType, data, ndim):
+    """Test that annotating with napari.types.<layer_type>Data works.
+
+    It expects a raw data format (like a numpy array) and will add a layer
+    of the corresponding type to the viewer.
+    """
+    from concurrent.futures import Future
+    from functools import partial
+
+    from qtpy.QtCore import QTimer
+
+    viewer = make_napari_viewer()
+    dtype = getattr(types, f'{LayerType.__name__}Data')
+
+    @magicgui
+    # where `dtype` is something like napari.types.ImageData
+    def add_data() -> Future[dtype]:  # type: ignore
+        future = Future()
+        # simulate something that isn't immediately ready when function returns
+        QTimer.singleShot(10, partial(future.set_result, data))
+        return future
+
+    viewer.window.add_dock_widget(add_data)
+
+    def _assert_stuff():
+        assert len(viewer.layers) == 1
+        assert isinstance(viewer.layers[0], LayerType)
+        assert viewer.layers[0].source.widget == add_data
+
+    add_data()
+    assert len(viewer.layers) == 0
+    QTimer.singleShot(50, _assert_stuff)
+    time.sleep(0.1)
```

Comment:
fwiw, played with the test a bit. I was curious about doing it without the sleep, and I also wondered about using the standard executor.

```python
    exec = ThreadPoolExecutor()

    @magicgui
    # where `dtype` is something like napari.types.ImageData
    def add_data() -> Future[dtype]:  # type: ignore
        def do_add_data():
            time.sleep(0.1)
            return data
        return exec.submit(do_add_data)

    viewer.window.add_dock_widget(add_data)

    def _assert_stuff():
        assert len(viewer.layers) == 1
        assert isinstance(viewer.layers[0], LayerType)
        assert viewer.layers[0].source.widget == add_data

    # wait for the add_data() future to resolve
    add_data()
    assert(len(viewer.layers) == 0)
    exec.shutdown(wait=True)
    QTimer.singleShot(0,
                      _assert_stuff)
```

The `singleShot` at the end is required so the event loop gets pumped. The `Sentry` gets alerted by the time `exec.shutdown()` is done, but the Qt event won't get processed until somebody interacts with the event queue.